corehq/apps/app_manager/suite_xml/utils.py

```diff
@@ -8,17 +8,33 @@
 
 
 def get_select_chain(app, module, include_self=True):
-    select_chain = [module] if include_self else []
+    return [
+        module
+        for (module, _) in get_select_chain_with_sessions(app, module, include_self=include_self)
+    ]
+
+
+def get_select_chain_with_sessions(app, module, include_self=True):
+    select_chain = [(module, 'case_id')] if include_self else []
     current_module = module
+    case_type = module.case_type
+    i = len(select_chain)
     while hasattr(current_module, 'parent_select') and current_module.parent_select.active:
+        is_other_relation = current_module.parent_select.relationship is None
         current_module = app.get_module_by_unique_id(
             current_module.parent_select.module_id,
             error=_("Case list used by parent child selection in '{}' not found").format(
                 current_module.default_name()),
         )
+        if is_other_relation and case_type == current_module.case_type:
+            session_var = 'case_id_' + case_type
+        else:
+            session_var = ('parent_' * i or 'case_') + 'id'
+        case_type = current_module.case_type
         if current_module in select_chain:
             raise SuiteValidationError("Circular reference in case hierarchy")
-        select_chain.append(current_module)
+        select_chain.append((current_module, session_var))
+        i += 1
```

Comment:
Just double-checking it's intentional that `i` start at `len(select_chain)` and then increase - I'd have expected `i` to go from `len(select_chain)` down to 0 (or the reverse, I'm not that clear on this code).

Reply:
Hey, yes, this is intentional.

Reply:
You can see that this way `get_select_chain_meta` gets the order correct since it reverses what this function sends.